File: bl/runners/subprocess_runner.py
"""
bl/runners/subprocess_runner.py — General subprocess command runner (C-08).

Executes arbitrary shell commands specified in the question's Test field.
Interprets exit codes and stdout patterns to determine verdict.

Handles questions with mode: subprocess

Test field syntax:
    {shell command}                   — run this command
    expect_exit: {N}                  — expected exit code (default 0)
    expect_stdout: {substring}        — stdout must contain this
    expect_not_stdout: {substring}    — stdout must NOT contain this
    timeout: {N}                      — timeout in seconds (default 30)

If the command outputs a JSON object with a "verdict" key, that verdict
takes precedence over exit-code interpretation.
"""

import json
import subprocess

from bl.config import cfg

# ---------------------------------------------------------------------------
# Directive prefixes — lines starting with these are config, not command
# ---------------------------------------------------------------------------

_DIRECTIVES = ("expect_exit:", "expect_stdout:", "expect_not_stdout:", "timeout:")
_CODE_FENCE = "```"


def _parse_subprocess_spec(test_field: str) -> dict:
    """Parse the Test field into a subprocess execution spec.

    Returns a dict with keys:
        command, expect_exit, expect_stdout, expect_not_stdout, timeout
    """
    spec = {
        "command": None,
        "expect_exit": 0,
        "expect_stdout": None,
        "expect_not_stdout": None,
        "timeout": 30,
    }

    command_lines = []

    for line in test_field.splitlines():
        stripped = line.strip()

        # Strip markdown code fences
        if stripped.startswith(_CODE_FENCE):
            continue

        if not stripped:
            continue

        low = stripped.lower()

        if low.startswith("expect_exit:"):
            try:
                spec["expect_exit"] = int(stripped.split(":", 1)[1].strip())
            except ValueError:
                pass
            continue

        if low.startswith("expect_stdout:"):
            spec["expect_stdout"] = stripped.split(":", 1)[1].strip()
            continue

        if low.startswith("expect_not_stdout:"):
            spec["expect_not_stdout"] = stripped.split(":", 1)[1].strip()
            continue

        if low.startswith("timeout:"):
            try:
                spec["timeout"] = int(stripped.split(":", 1)[1].strip())
            except ValueError:
                pass
            continue

        # Non-directive, non-empty line is part of the command
        command_lines.append(stripped)

    if command_lines:
        spec["command"] = " && ".join(command_lines)

    return spec


# ---------------------------------------------------------------------------
# Main runner
# ---------------------------------------------------------------------------


def run_subprocess(question: dict) -> dict:
    """Execute shell command from Test field and return verdict envelope."""
    test_field = question.get("test", "") or question.get("Test", "")
    spec = _parse_subprocess_spec(test_field)

    command = spec["command"]
    expect_exit = spec["expect_exit"]
    expect_stdout = spec["expect_stdout"]
    expect_not_stdout = spec["expect_not_stdout"]
    timeout = spec["timeout"]

    if not command:
        return {
            "verdict": "INCONCLUSIVE",
            "summary": "No command found in Test field",
            "data": {"test_field": test_field[:200]},
            "details": "The Test field contained no executable command lines. "
            "Add a shell command before any expect_*/timeout: directives.",
        }

    # Execute
    returncode = -1
    stdout = ""
    stderr = ""

    try:
        result = subprocess.run(
            command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=timeout,
            cwd=str(cfg.autosearch_root),
        )
        returncode = result.returncode
        stdout = result.stdout or ""
        stderr = result.stderr or ""

    except subprocess.TimeoutExpired:
        return {
            "verdict": "FAILURE",
            "summary": f"Command timed out after {timeout}s: {command[:80]}",
            "data": {
                "command": command,
                "returncode": None,
                "exit_expected": expect_exit,
                "stdout_preview": "",
                "stderr_preview": "",
                "failure_type": "timeout",
            },
            "details": f"Command: {command}\nTimeout: {timeout}s exceeded",
            "failure_type": "timeout",
        }
    except Exception as exc:
        return {
            "verdict": "INCONCLUSIVE",
            "summary": f"Failed to run command: {exc}",
            "data": {"command": command, "error": str(exc)},
            "details": f"Command: {command}\nError: {exc}",
        }

    # Check if stdout is a JSON verdict envelope
    stdout_stripped = stdout.strip()
    if stdout_stripped.startswith("{"):
        try:
            parsed = json.loads(stdout_stripped)
            if "verdict" in parsed:
                parsed.setdefault(
                    "summary", f"exit={returncode}: {stdout_stripped[:100]}"
                )
                parsed.setdefault(
                    "data",
                    {
                        "command": command,
                        "returncode": returncode,
                        "exit_expected": expect_exit,
                        "stdout_preview": stdout[:500],
                        "stderr_preview": stderr[:200],
                    },
                )
                parsed.setdefault(
                    "details",
                    (
                        f"Command: {command}\n"
                        f"Exit: {returncode} (expected {expect_exit})\n"
                        f"Stdout:\n{stdout[:800]}\n"
                        f"Stderr:\n{stderr[:400]}"
                    ),
                )
                return parsed
        except json.JSONDecodeError:
            pass

    # Determine verdict from exit code and pattern checks
    verdict = "HEALTHY"
    failure_reasons = []

    if returncode != expect_exit:
        verdict = "FAILURE"
        failure_reasons.append(f"exit code {returncode} != expected {expect_exit}")

    if expect_stdout is not None and expect_stdout not in stdout:
        verdict = "FAILURE"
        failure_reasons.append(f"expected stdout substring '{expect_stdout}' not found")

    if expect_not_stdout is not None and expect_not_stdout in stdout:
        verdict = "FAILURE"
        failure_reasons.append(
            f"forbidden stdout substring '{expect_not_stdout}' was found"
        )

    one_line = (stdout[:100] or stderr[:100]).strip().replace("\n", " ")
    summary = f"exit={returncode}: {one_line}"
    if failure_reasons:
        summary += " | " + "; ".join(failure_reasons)

    return {
        "verdict": verdict,
        "summary": summary,
        "data": {
            "command": command,
            "returncode": returncode,
            "exit_expected": expect_exit,
            "stdout_preview": stdout[:500],
            "stderr_preview": stderr[:200],
        },
        "details": (
            f"Command: {command}\n"
            f"Exit: {returncode} (expected {expect_exit})\n"
            f"Stdout:\n{stdout[:800]}\n"
            f"Stderr:\n{stderr[:400]}"
        ),
    }

File: masonry/src/engine/runners/subprocess-runner.js
'use strict';
// engine/runners/subprocess-runner.js — General subprocess command runner.
//
// Port of bl/runners/subprocess_runner.py to Node.js.

const { execSync } = require('child_process');
const { cfg } = require('../config');

const _DIRECTIVES = ['expect_exit:', 'expect_stdout:', 'expect_not_stdout:', 'timeout:'];
const _CODE_FENCE = '```';

function _parseSubprocessSpec(testField) {
  const spec = {
    command: null,
    expectExit: 0,
    expectStdout: null,
    expectNotStdout: null,
    timeout: 30,
  };

  const commandLines = [];

  for (const line of testField.split('\n')) {
    const stripped = line.trim();
    if (stripped.startsWith(_CODE_FENCE)) continue;
    if (!stripped) continue;

    const low = stripped.toLowerCase();

    if (low.startsWith('expect_exit:')) {
      const val = parseInt(stripped.slice(stripped.indexOf(':') + 1).trim(), 10);
      if (!isNaN(val)) spec.expectExit = val;
      continue;
    }
    if (low.startsWith('expect_stdout:')) {
      spec.expectStdout = stripped.slice(stripped.indexOf(':') + 1).trim();
      continue;
    }
    if (low.startsWith('expect_not_stdout:')) {
      spec.expectNotStdout = stripped.slice(stripped.indexOf(':') + 1).trim();
      continue;
    }
    if (low.startsWith('timeout:')) {
      const val = parseInt(stripped.slice(stripped.indexOf(':') + 1).trim(), 10);
      if (!isNaN(val)) spec.timeout = val;
      continue;
    }

    commandLines.push(stripped);
  }

  if (commandLines.length) {
    spec.command = commandLines.join(' && ');
  }

  return spec;
}

function _buildVerdict({ command, returncode, stdout, stderr, expectExit, expectStdout, expectNotStdout }) {
  // Check if stdout is a JSON verdict envelope
  const stdoutTrimmed = stdout.trim();
  if (stdoutTrimmed.startsWith('{')) {
    try {
      const parsed = JSON.parse(stdoutTrimmed);
      if (parsed.verdict) {
        if (!parsed.summary) parsed.summary = `exit=${returncode}: ${stdoutTrimmed.slice(0, 100)}`;
        if (!parsed.data) parsed.data = { command, returncode, exit_expected: expectExit, stdout_preview: stdout.slice(0, 500), stderr_preview: stderr.slice(0, 200) };
        if (!parsed.details) parsed.details = `Command: ${command}\nExit: ${returncode} (expected ${expectExit})\nStdout:\n${stdout.slice(0, 800)}\nStderr:\n${stderr.slice(0, 400)}`;
        return parsed;
      }
    } catch {
      // not valid JSON
    }
  }

  let verdict = 'HEALTHY';
  const failureReasons = [];

  if (returncode !== expectExit) {
    verdict = 'FAILURE';
    failureReasons.push(`exit code ${returncode} != expected ${expectExit}`);
  }

  if (expectStdout !== null && !stdout.includes(expectStdout)) {
    verdict = 'FAILURE';
    failureReasons.push(`expected stdout substring '${expectStdout}' not found`);
  }

  if (expectNotStdout !== null && stdout.includes(expectNotStdout)) {
    verdict = 'FAILURE';
    failureReasons.push(`forbidden stdout substring '${expectNotStdout}' was found`);
  }

  const oneLine = (stdout.slice(0, 100) || stderr.slice(0, 100)).trim().replace(/\n/g, ' ');
  let summary = `exit=${returncode}: ${oneLine}`;
  if (failureReasons.length) {
    summary += ' | ' + failureReasons.join('; ');
  }

  return {
    verdict,
    summary,
    data: {
      command,
      returncode,
      exit_expected: expectExit,
      stdout_preview: stdout.slice(0, 500),
      stderr_preview: stderr.slice(0, 200),
    },
    details: `Command: ${command}\nExit: ${returncode} (expected ${expectExit})\nStdout:\n${stdout.slice(0, 800)}\nStderr:\n${stderr.slice(0, 400)}`,
  };
}

function runSubprocess(question) {
  const testField = question.test || question.Test || '';
  const spec = _parseSubprocessSpec(testField);

  if (!spec.command) {
    return {
      verdict: 'INCONCLUSIVE',
      summary: 'No command found in Test field',
      data: { test_field: testField.slice(0, 200) },
      details: 'The Test field contained no executable command lines.',
    };
  }

  try {
    const stdout = execSync(spec.command, {
      encoding: 'utf8',
      timeout: spec.timeout * 1000,
      cwd: cfg.autosearchRoot,
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    return _buildVerdict({
      command: spec.command,
      returncode: 0,
      stdout: stdout || '',
      stderr: '',
      expectExit: spec.expectExit,
      expectStdout: spec.expectStdout,
      expectNotStdout: spec.expectNotStdout,
    });
  } catch (err) {
    if (err.killed) {
      return {
        verdict: 'FAILURE',
        summary: `Command timed out after ${spec.timeout}s: ${spec.command.slice(0, 80)}`,
        data: { command: spec.command, returncode: null, failure_type: 'timeout' },
        details: `Command: ${spec.command}\nTimeout: ${spec.timeout}s exceeded`,
        failure_type: 'timeout',
      };
    }
    if (err.status !== undefined) {
      return _buildVerdict({
        command: spec.command,
        returncode: err.status,
        stdout: (err.stdout || '').toString(),
        stderr: (err.stderr || '').toString(),
        expectExit: spec.expectExit,
        expectStdout: spec.expectStdout,
        expectNotStdout: spec.expectNotStdout,
      });
    }
    return {
      verdict: 'INCONCLUSIVE',
      summary: `Failed to run command: ${err.message}`,
      data: { command: spec.command, error: err.message },
      details: `Command: ${spec.command}\nError: ${err.message}`,
    };
  }
}

module.exports = {
  runSubprocess,
  _parseSubprocessSpec,
  _buildVerdict,
};
